lib/datalog/roletester/src/role_tester.rs

```diff
@@ -0,0 +1,220 @@
+            runtime.extend(&[HasTrait(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        } else if pred.name == types::facts::PredicateType::RoleAllowsLogin as i32 {
+            runtime.extend(&[RoleAllowsLogin(pred.atoms[0], pred.atoms[1])]);
+        } else if pred.name == types::facts::PredicateType::RoleDeniesLogin as i32 {
+            runtime.extend(&[RoleDeniesLogin(pred.atoms[0], pred.atoms[1])]);
+        } else if pred.name == types::facts::PredicateType::RoleAllowsNodeLabel as i32 {
+            runtime.extend(&[RoleAllowsNodeLabel(
+                pred.atoms[0],
+                pred.atoms[1],
+                pred.atoms[2],
+            )]);
+        } else if pred.name == types::facts::PredicateType::RoleDeniesNodeLabel as i32 {
+            runtime.extend(&[RoleDeniesNodeLabel(
+                pred.atoms[0],
+                pred.atoms[1],
+                pred.atoms[2],
+            )]);
+        } else if pred.name == types::facts::PredicateType::NodeHasLabel as i32 {
+            runtime.extend(&[NodeHasLabel(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        }
+    }
+
+    let (accesses, deny_accesses, deny_logins) = runtime.run();
+    let mut predicates = vec![];
+    predicates.extend(
+        accesses
+            .into_iter()
+            .map(|HasAccess(a, b, c, d)| types::facts::Predicate {
+                name: types::facts::PredicateType::HasAccess as i32,
+                atoms: vec![a, b, c, d],
+            }),
+    );
+    predicates.extend(
+        deny_accesses
+            .into_iter()
+            .map(|DenyAccess(a, b, c, d)| types::facts::Predicate {
+                name: types::facts::PredicateType::DenyAccess as i32,
+                atoms: vec![a, b, c, d],
+            })
+    );
+    predicates.extend(
+        deny_logins
+            .into_iter()
+            .map(|DenyLogins(a, b, c)| types::facts::Predicate {
+                name: types::facts::PredicateType::DenyLogins as i32,
+                atoms: vec![a, b, c],
+            })
+    );
+
+    let idb = types::Facts {
+        predicates,
+    };
+
+    let mut buf = Vec::with_capacity(idb.encoded_len());
+        if let Err(e) = idb.encode(&mut buf) {
+            let err_bytes = e.to_string().into_bytes();
+            unsafe {
+                ptr::copy(&(*err_bytes)[0], output, err_bytes.len());
+            }
+            return Box::into_raw(Box::new(Status{
+                output_length: err_bytes.len(),
+                error: -1,
+            }))
+        }
+    if buf.is_empty() || buf.len() > output_len {
+        return Box::into_raw(Box::new(Status{
+            output_length: buf.len(),
+            error: 0,
+        }))
+    }
+    // We can't guarantee the memory regions are non-overlapping, but we only need to copy the data to output so
+    // it is able to be presented on the Go end. The necessary checks for length is done before this call.
```

Comment:
IIUC `buf` and `output` are non-overlapping.  Looks to me like they result from separate allocations, which are always non-overlapping.  The only reason that you should ever end up with non-overlapping regions is if you are moving data around within a single contiguous buffer.

Reply:
Right, that makes sense.